Getting started with trixi

ArseniyKholod opened the pull request:
First version of 0th tutorial. (not completed)

sloede commented:
Hi @ArseniyKholod! Thanks for getting started with this. At the moment, all your files live in an extra directory, however, we want the tutorial to be integrated in our general documentation infrastructure. Could you thus please move the Literate files (i.e., the "normal" .jl file to the `docs/literate/src/files` directory (where also the other tutorials exist), and then add the filename and title (for the navigation bar) to 
https://github.com/ArseniyKholod/Trixi.jl/blob/cb61c3a4ed896e62750408ef5de1f97e47c85ceb/docs/make.jl#L50-L71
?

This should set up for the tutorial to be build automatically when the documentation is built. Then it becomes much easier for me to evaluate the contents of the tutorial :-)

ArseniyKholod commented:
Hi @sloede!
I moved file to docs/literate/src/files and added named changes to make.jl.
But I'm not sure, that it is possible to update https://trixi-framework.github.io/Trixi.jl/stable/ from my forked Trixi copy.

ArseniyKholod commented:
Now should be no troubles with compilation of all tutorials including new one (0th)

sloede commented on docs/make.jl:
```diff
@@ -48,6 +48,8 @@ end
 #               "subtitle 2" => ("folder 2", "filename 2.jl")]
 files = [
+    # Topic: introduction
+    "Getting started with Trixi" => "getting_started_with_Trixi.jl",
```
```suggestion
    "Getting started" => "getting_started.jl",
```
I think it's redundant to refer to Trixi.jl here. Also, I'd rename the file accordingly for consistency

sloede commented on docs/literate/src/files/getting_started_with_Trixi.jl:
```diff
@@ -0,0 +1,287 @@
+#src # Getting started with Trixi.jl
```
```suggestion
#src # Getting started
```

sloede commented on docs/literate/src/files/getting_started_with_Trixi.jl:
Rename file to `getting_started.jl`

sloede commented on docs/literate/src/files/getting_started_with_Trixi.jl:
```diff
@@ -0,0 +1,287 @@
+#src # Getting started with Trixi.jl
+
+# This tutorial is intended for beginners in Julia and Trixi.jl.
+# After reading it, you will install Julia and Trixi on your PC and will be able to download files
+# from Trixi github, execute and edit them.
```
```suggestion
# After reading it, you will know how to install Julia and Trixi.jl on your computer,
# and you will be able to download setup files from our GitHub repository, modify them,
# and run simulations.
```

sloede commented on docs/literate/src/files/getting_started_with_Trixi.jl:
```diff
@@ -0,0 +1,287 @@
+# This tutorial is intended for beginners in Julia and Trixi.jl.
+# After reading it, you will install Julia and Trixi on your PC and will be able to download files
+# from Trixi github, execute and edit them.
+# **Trixi.jl** is a numerical simulation framework for hyperbolic conservation laws 
+# written in [`Julia`](https://julialang.org/).
```
```suggestion
# Trixi.jl is a numerical simulation framework for conservation laws and
# is written in the [Julia programming language](https://julialang.org/).
```
Maybe put this sentence as the first one of the paragraph?

sloede commented on docs/literate/src/files/getting_started_with_Trixi.jl:
```diff
@@ -0,0 +1,287 @@
+# from Trixi github, execute and edit them.
+# **Trixi.jl** is a numerical simulation framework for hyperbolic conservation laws 
+# written in [`Julia`](https://julialang.org/).
+# This means that Julia have to be installed on a PC to work with Trixi. 
```
```suggestion
```
I think this information is redundant

sloede commented on docs/literate/src/files/getting_started_with_Trixi.jl:
```diff
@@ -0,0 +1,287 @@
+
+# ## Julia installation
+
+# Trixi works with the current stable Julia release. More information about Julia support can be
```
```suggestion
# Trixi.jl works with the current stable Julia release. More information about Julia support can be
```
Please always use "Trixi.jl" (i.e., with the ".jl") when referring to Trixi.jl. Please fix this also in all other locations where this might be used wrongly in this PR.

sloede commented on docs/literate/src/files/getting_started_with_Trixi.jl:
```diff
@@ -0,0 +1,287 @@
+# ## Julia installation
+
+# Trixi works with the current stable Julia release. More information about Julia support can be
+# found in [`README`](https://github.com/trixi-framework/Trixi.jl#installation).
```
```suggestion
# found in the [`README.md`](https://github.com/trixi-framework/Trixi.jl#installation) file.
```

sloede commented on docs/literate/src/files/getting_started_with_Trixi.jl:
```diff
@@ -0,0 +1,287 @@
+
+# Trixi works with the current stable Julia release. More information about Julia support can be
+# found in [`README`](https://github.com/trixi-framework/Trixi.jl#installation).
+# The most fully explained installation process can be found in
+# [`Julia installation instruction`](https://julialang.org/downloads/platform/).
+# But you can follow also our short installation instruction.
```
```suggestion
# A detailed description of the installation process can be found in the
# [Julia installation instructions](https://julialang.org/downloads/platform/).
# But you can follow also our short installation guidelines below.
```

sloede commented on docs/literate/src/files/getting_started_with_Trixi.jl:
```diff
@@ -0,0 +1,287 @@
+
+# ### Windows
+
+# - Download Julia [`installer`](https://julialang.org/downloads/) for Windows. Make sure 
```
```suggestion
# - Download Julia installer for Windows from https://julialang.org/downloads/. Make sure 
```

sloede commented on docs/literate/src/files/getting_started_with_Trixi.jl:
```diff
@@ -0,0 +1,287 @@
+# - Open the downloaded installer.
+# - Paste an installation directory path or find it using a file manager (select `Browse`).
+# - Select `Next`.
+# - Check the `Add Julia to PATH` to add Julia to the Environment Variables. 
+#   This makes possible to run Julia using Terminal from any directory only typing `julia`.
+# - Select `Next`, then Julia will be insalled.
```
```suggestion
# - Check the `Add Julia to PATH` option to add Julia to the environment variables. 
#   This makes it possible to run Julia in the terminal from any directory by only typing `julia`.
# - Select `Next`, then Julia will be installed.
```

sloede commented on docs/literate/src/files/getting_started_with_Trixi.jl:
```diff
@@ -0,0 +1,287 @@
+# - Enter `cmd` in opened window.
+# - Enter in a terminal `julia`. 
+
+# Then Julia will be invoked. To close Julia enter `exit()`. 
```
```suggestion
# Then Julia will be invoked. To close Julia enter `exit()` or press `Ctrl+d`. 
```

sloede commented on docs/literate/src/files/getting_started_with_Trixi.jl:
```diff
@@ -0,0 +1,287 @@
+# - Select `Next`, then Julia will be insalled.
+
+# Now you can verify, that Julia is installed:
+# - Type `Win+R` on a keyboard.
```
```suggestion
# - Press `Win+R` on a keyboard.
```
To me, "type" is ambiguous (it could also mean that they should literally type "W i n + R")

sloede commented on docs/literate/src/files/getting_started_with_Trixi.jl:
```diff
@@ -0,0 +1,287 @@
+
+# ### Linux
+
+# - Open a Terminal and navigate (using `cd`) to the directory, where you want to save Julia.
```
```suggestion
# - Open a terminal and navigate (using `cd`) to the directory, where you want to store Julia.
```

sloede commented on docs/literate/src/files/getting_started_with_Trixi.jl:
```diff
@@ -0,0 +1,287 @@
+# ### Linux
+
+# - Open a Terminal and navigate (using `cd`) to the directory, where you want to save Julia.
+#   Or you can open a file manager, find this directory, right-click inside and 
+#   choose `Open Terminal Here`.
```
```suggestion
```
I would omit this since it may differ from Linux distro to Linux distro

sloede commented on docs/literate/src/files/getting_started_with_Trixi.jl:
`````diff
@@ -0,0 +1,287 @@
+# - Open a Terminal and navigate (using `cd`) to the directory, where you want to save Julia.
+#   Or you can open a file manager, find this directory, right-click inside and 
+#   choose `Open Terminal Here`.
+# - To install Julia execute the following commands in a Terminal:
+#   ````
+#   wget https://julialang-s3.julialang.org/bin/linux/x64/1.8/julia-1.8.5-linux-x86_64.tar.gz
+#   tar zxvf julia-1.8.5-linux-x86_64.tar.gz
`````
This is specific for Julia v1.8.5. Can you make this more generic (maybe without giving the links explicitly here), but something like
> Get the link to the latest Julia release, then download by executing
```
wget https://julialang-s3.julialang.org/bin/linux/... # your link goes here
```
> Now unpack the downloaded file with
```
tar xf julia-....tar.gz # your filename goes here
```

?

sloede commented on docs/literate/src/files/getting_started_with_Trixi.jl:
`````diff
@@ -0,0 +1,287 @@
+#   wget https://julialang-s3.julialang.org/bin/linux/x64/1.8/julia-1.8.5-linux-x86_64.tar.gz
+#   tar zxvf julia-1.8.5-linux-x86_64.tar.gz
+#   ````
+#   Now you can verify that Julia is installed entering `julia` command in a Terminal.
`````
I think this will not work, right? They have to call something like `julia-1.8.5/bin/julia`. Maybe you can update this section?

sloede commented on docs/literate/src/files/getting_started_with_Trixi.jl:
`````diff
@@ -0,0 +1,287 @@
+#   ````
+#   Now you can verify that Julia is installed entering `julia` command in a Terminal.
+
+# Then Julia will be invoked. To close Julia enter `exit()`.
`````
```suggestion
# Then Julia will be invoked. To close Julia, enter `exit()` or press `Ctrl+d`.
```

sloede commented on docs/literate/src/files/getting_started_with_Trixi.jl:
```diff
@@ -0,0 +1,287 @@
+
+# ## Trixi installation
+
+# Trixi and its related tools are registered Julia packages. So installation of them is
+# running inside Julia. For appropriate work of Trixi you need to install 
+# [`Trixi`](https://github.com/trixi-framework/Trixi.jl),
+# [`Trixi2Vtk`](https://github.com/trixi-framework/Trixi2Vtk.jl),
+# [`OrdinaryDiffEq`](https://github.com/SciML/OrdinaryDiffEq.jl) and 
+# [`Plots`](https://github.com/JuliaPlots/Plots.jl).
```
```suggestion
# Trixi and its related tools are registered Julia packages, thus their installation
# happens inside Julia.
# For a smooth workflow experience with Trixi.jl, you need to install 
# [Trixi.jl](https://github.com/trixi-framework/Trixi.jl),
# [OrdinaryDiffEq.jl](https://github.com/SciML/OrdinaryDiffEq.jl) and 
# [Plots.jl](https://github.com/JuliaPlots/Plots.jl).
```
Some rewording. I also removed Trixi2Vtk, since ParaView is not mentioned in this tutorial at all, so it is not strictly necessary here I believe.

sloede commented on docs/literate/src/files/getting_started_with_Trixi.jl:
```diff
@@ -0,0 +1,287 @@
+# [`OrdinaryDiffEq`](https://github.com/SciML/OrdinaryDiffEq.jl) and 
+# [`Plots`](https://github.com/JuliaPlots/Plots.jl).
+
+# - Open a Terminal.
```
```suggestion
# - Open a terminal.
```

sloede commented on docs/literate/src/files/getting_started_with_Trixi.jl:
```diff
@@ -0,0 +1,287 @@
+# [`Plots`](https://github.com/JuliaPlots/Plots.jl).
+
+# - Open a Terminal.
+# - Invoke Julia executing `julia`.
```
Again, this only works at the moment for Windows, since you have not explained how to add `julia` to the PATH on Linux/macOS. You don't have to, but maybe you can then modify this line accordingly?

ArseniyKholod replied:
I added in the part with installing Julia on Linux a link to official explanation of Julia-team, how to add `julia` to PATH. Without adding I have to change `julia` into `path/to/<Julia directory>` in part with cloning Trixi.jl, that will disable possibility to just copy and paste commands in a terminal. Alternatively, I can explain exactly the way, how to setup `julia` to the PATH on Linux, without referencing to another guide. But problem with MacOs remains, I can only copy commands or guides from other authors from internet, but without guarantee of course, that it will work. (I've no MacOs at hand)

sloede replied:
Have you tried [juliaup](https://github.com/JuliaLang/juliaup)? Maybe this could be an alternative to explain how to install Julia, as it seems to set things like PATH automatically. It will install the latest Julia release and you can always provide a link to the manual downloadable files as a fallback.

sloede commented on docs/literate/src/files/getting_started_with_Trixi.jl:
`````diff
@@ -0,0 +1,287 @@
+# - Execute following commands:
+#   ````
+#   import Pkg
+#   Pkg.add(["Trixi", "Trixi2Vtk", "OrdinaryDiffEq", "Plots"])
`````
```suggestion
#   Pkg.add(["OrdinaryDiffEq", "Plots", "Trixi"])
```
No Trixi2Vtk

sloede commented on docs/literate/src/files/getting_started_with_Trixi.jl:
`````diff
@@ -0,0 +1,287 @@
+#   ````
+
+# Now you have installed all these 
+# packages. [`Trixi2Vtk`](https://github.com/trixi-framework/Trixi2Vtk.jl) is a visualization
+# tool, [`OrdinaryDiffEq`](https://github.com/SciML/OrdinaryDiffEq.jl) provides time integration schemes
+# used by Trixi and [`Plots`](https://github.com/JuliaPlots/Plots.jl) can be used to directly
+# visualize Trixi results from the Julia REPL.
`````
```suggestion
# packages. [OrdinaryDiffEq.jl](https://github.com/SciML/OrdinaryDiffEq.jl) provides time integration schemes
# used by Trixi.jl and [Plots.jl](https://github.com/JuliaPlots/Plots.jl) can be used to directly
# visualize Trixi.jl results from the Julia REPL.
```

sloede commented on docs/literate/src/files/getting_started_with_Trixi.jl:
```diff
@@ -0,0 +1,287 @@
+
+# ## Usage
+
+# ### Files execution
```
```suggestion
# ### Running a simulation
```

sloede commented on docs/literate/src/files/getting_started_with_Trixi.jl:
```diff
@@ -0,0 +1,287 @@
+
+# ### Files execution
+
+# Trixi has a big set
+# of [`examples`](https://github.com/trixi-framework/Trixi.jl/tree/main/examples), that can be taken
+# as basis for your future investigations.
```
```suggestion
# To get you started, Trixi.jl has a large set
# of [example setups](https://github.com/trixi-framework/Trixi.jl/tree/main/examples), that can be taken
# as a basis for your future investigations.
# In Trixi.jl, we call these setup files "elixirs", since they contain Julia code that
# takes parts of Trixi.jl and combines them into something new.
```

sloede commented on docs/literate/src/files/getting_started_with_Trixi.jl:
```diff
@@ -0,0 +1,287 @@
+# of [`examples`](https://github.com/trixi-framework/Trixi.jl/tree/main/examples), that can be taken
+# as basis for your future investigations.
+
+# Now execute one of them using 
+# [`trixi_include(...)`](https://trixi-framework.github.io/Trixi.jl/stable/reference-trixi/#Trixi.trixi_include-Tuple{Module,%20AbstractString})
+# function. `trixi_include(...)` expects
```
```suggestion
# Now execute one of the examples using the [`trixi_include`](@ref)
# function. `trixi_include(...)` expects
```

sloede commented on docs/literate/src/files/getting_started_with_Trixi.jl:
```diff
@@ -0,0 +1,287 @@
+# Now execute one of them using 
+# [`trixi_include(...)`](https://trixi-framework.github.io/Trixi.jl/stable/reference-trixi/#Trixi.trixi_include-Tuple{Module,%20AbstractString})
+# function. `trixi_include(...)` expects
+# a single string argument with a path to a text file containing Julia code.
```
```suggestion
# a single string argument with a path to a file containing Julia code.
```

sloede commented on docs/literate/src/files/getting_started_with_Trixi.jl:
```diff
@@ -0,0 +1,287 @@
+# [`trixi_include(...)`](https://trixi-framework.github.io/Trixi.jl/stable/reference-trixi/#Trixi.trixi_include-Tuple{Module,%20AbstractString})
+# function. `trixi_include(...)` expects
+# a single string argument with a path to a text file containing Julia code.
+# `joinpath(...)` join a path components into a full path. 
+# [`examples_dir()`](https://trixi-framework.github.io/Trixi.jl/stable/reference-trixi/#Trixi.examples_dir-Tuple{})
+# returns a path to the
+# [`examples`](https://github.com/trixi-framework/Trixi.jl/tree/main/examples) folder.
```
```suggestion
# `joinpath(...)` join a path components into a full path. 
# The [`examples_dir`](@ref) function returns a path to the
# [`examples`](https://github.com/trixi-framework/Trixi.jl/tree/main/examples) folder
# that has been locally downloaded while installing Trixi.jl.
```

sloede commented on docs/literate/src/files/getting_started_with_Trixi.jl:
```diff
@@ -0,0 +1,287 @@
+# returns a path to the
+# [`examples`](https://github.com/trixi-framework/Trixi.jl/tree/main/examples) folder.
+
+# Let's execute a short two-dimensional problem setup. Which approximates the solution of
+# [`compressible Euler equations in 2D for an ideal gas`](https://trixi-framework.github.io/Trixi.jl/stable/reference-trixi/#Trixi.CompressibleEulerEquations2D)
+# with
+# [`weak blast wave initial condition`](https://trixi-framework.github.io/Trixi.jl/stable/reference-trixi/#Trixi.initial_condition_weak_blast_wave-Tuple{Any,%20Any,%20CompressibleEulerEquations2D})
```
```suggestion
# Let's execute a short two-dimensional problem setup. It approximates the solution of
# the compressible Euler equations in 2D for an ideal gas ([`CompressibleEulerEquations2D`](@ref))
# with a weak blast wave as the initial condition.
```

sloede commented on docs/literate/src/files/getting_started_with_Trixi.jl:
```diff
@@ -0,0 +1,287 @@
+# [`weak blast wave initial condition`](https://trixi-framework.github.io/Trixi.jl/stable/reference-trixi/#Trixi.initial_condition_weak_blast_wave-Tuple{Any,%20Any,%20CompressibleEulerEquations2D})
+
+# Invoke Julia in a Terminal.
+# And execute following code. (*Remark:* you can ignore all arguments of trixi_include() except
+# path to the file) 
```
Why _are_ these additional parts there, .e.g, the `module` and the callback stuff?

ArseniyKholod replied:
'module' is here to enable using 'sol' variable for 'plot' function, otherwise error appears, that 'sol' is undefined.
'callbacks' are used to prevent from saving the solution, problem is in [this part of hohqmesh_tutorial](https://trixi-framework.github.io/Trixi.jl/stable/tutorials/hohqmesh_tutorial/#Running-and-visualizing-an-unstructured-simulation). Here converting actions on .h5 files are performed, and error because of inconsistent of mesh files appears. ('trixi_include()' is used only in hohqmesh_tutorial and in new one, so there was no problem before)

sloede replied:
Is it possible to clean the `out` directory after this tutorial is run? this should fix the issue with the HOHQMesh tutorial. Alternatively, we can try to clean the `out` directory before the HOHQMesh tutorial is run.

ArseniyKholod replied:
I applied first option. It works, thank you!

sloede commented on docs/literate/src/files/getting_started.jl:
`````diff
@@ -0,0 +1,288 @@
+
+# - Open the downloaded file with notepad or any other text editor.
+# - And go to the 9th line with following code:
+#   ````
+#   initial_condition = initial_condition_weak_blast_wave
+#   ````
`````
~~~suggestion
#   ```julia
#   initial_condition = initial_condition_weak_blast_wave
#   ```
~~~
This should apply Julia formatting to the source code. Please also do this for all other Julia source code snippets that you include in markdown

sloede commented on docs/literate/src/files/getting_started.jl:
```diff
@@ -0,0 +1,288 @@
+
+trixi_include(@__MODULE__,joinpath(examples_dir(), "tree_2d_dgsem", "elixir_euler_ec.jl"), #hide #md
+  callbacks=CallbackSet(StepsizeCallback(cfl=1.0)), initial_condition=initial_condition) #hide #md
+plot(sol) #hide #md
```
The solution looks very ugly because of the weird choices for the axis limits by Plots.jl:
<img width="664" alt="image" src="https://user-images.githubusercontent.com/3637659/235155989-8bd97586-7968-4cc8-9fe1-b4b75b03c2a3.png">

Can you try to fix them please? You might need to use the `PlotData2D` approach and do something like this
```julia
pd = PlotData2D(sol)
p1 = plot(pd["rho"])
p2 = plot(pd["v1"], clim=(0.05, 0.25))
p3 = plot(pd["v2"], clim=(0.05, 0.25))
p4 = plot(pd["p"], clim=(10, 30))
plot(p1, p2, p3, p4)
```
which should give you something like
<img width="588" alt="image" src="https://user-images.githubusercontent.com/3637659/235163764-5774665f-c276-4a7a-90b6-2765a3b748d8.png">

sloede commented on docs/literate/src/files/getting_started.jl:
```diff
@@ -0,0 +1,288 @@
+  callbacks=CallbackSet(StepsizeCallback(cfl=1.0)), initial_condition=initial_condition) #hide #md
+plot(sol) #hide #md
+
+# Feel free to add
+# changes into `initial_condition` to observe different solutions.
+
+# Now you are able to download, edit and execute Trixi.jl code.
```
```suggestion
# Feel free to make further changes to the initial condition to observe different solutions.

# Now you are able to download, modify and execute simulation setups for Trixi.jl.
```

sloede commented on docs/literate/src/files/getting_started.jl:
```diff
@@ -0,0 +1,288 @@
+
+# Now you are able to download, edit and execute Trixi.jl code.
+
+# ## Next steps
```
```suggestion
# ## Next steps: changing Trixi.jl itself
```

sloede commented on docs/literate/src/files/getting_started.jl:
```diff
@@ -0,0 +1,288 @@
+
+# #### Windows
+
+# If you are using Windows OS, you can clone Trixi.jl directory using a Github Desktop.
+# - If you haven't any github account yet, you have to create it on
+#   the [Github website](https://github.com/join).
+# - Download and install [Github Desktop](https://desktop.github.com/) and then login into
+#   your account.
+# - Open an installed Github Desktop, press `Ctrl+Shift+O`.
+# - In opened window paste `trixi-framework/Trixi.jl` and choose path to the folder, where you want
+#   to save Trixi.jl. Then click `Clone` and Trixi.jl will be cloned to your computer. 
```
```suggestion
# If you are using Windows, you can clone Trixi.jl by using the GitHub Desktop tool:
# - If you do not have a GitHub account yet, create it on
#   the [GitHub website](https://github.com/join).
# - Download and install [GitHub Desktop](https://desktop.github.com/) and then log in into
#   your account.
# - Open GitHub Desktop, press `Ctrl+Shift+O`.
# - In the opened window, paste `trixi-framework/Trixi.jl` and choose the path to the folder where you want
#   to save Trixi.jl. Then click `Clone` and Trixi.jl will be cloned to your computer. 
```

sloede commented on docs/literate/src/files/getting_started.jl:
```diff
@@ -0,0 +1,288 @@
+# - In opened window paste `trixi-framework/Trixi.jl` and choose path to the folder, where you want
+#   to save Trixi.jl. Then click `Clone` and Trixi.jl will be cloned to your computer. 
+
+# Now you cloned Trixi.jl and only need to add Trixi.jl packages to Julia.
```
```suggestion
# Now you cloned Trixi.jl and only need to tell Julia to use the local clone as the package sources:
```

sloede commented on docs/literate/src/files/getting_started.jl:
`````diff
@@ -0,0 +1,288 @@
+
+# Now you cloned Trixi.jl and only need to add Trixi.jl packages to Julia.
+# - Open a terminal using `Win+R` and `cmd`. Navigate to the folder with cloned Trixi.jl using `cd`.
+# - Create new directory and start Julia with the `--project` flag set to your local Trixi.jl clone.
+#   ````
+#   mkdir run 
+#   cd run
+#   julia --project=.
+#   ````
`````
~~~suggestion
# - Create new directory `run`, enter it, and start Julia with the `--project=.` flag:
#   ```shell
#   mkdir run 
#   cd run
#   julia --project=.
#   ```
~~~

sloede commented on docs/literate/src/files/getting_started.jl:
`````diff
@@ -0,0 +1,288 @@
+#   cd run
+#   julia --project=.
+#   ````
+# - Run following commands in Julia REPL:
+#   ````
+#   using Pkg; Pkg.develop(PackageSpec(path="..")) # Install local Trixi.jl clone
+#   Pkg.add(["OrdinaryDiffEq", "Trixi2Vtk", "Plots"])  # Install additional packages
+#   ````
`````
```suggestion
# - Now run the following commands to install all relevant packages:
#   ````
#   using Pkg; Pkg.develop(PackageSpec(path="..")) # Install local Trixi.jl clone
#   Pkg.add(["OrdinaryDiffEq", "Plots"])  # Install additional packages
#   ````
```

sloede commented on docs/literate/src/files/getting_started.jl:
`````diff
@@ -0,0 +1,288 @@
+# You can clone Trixi.jl to your computer executing following commands:
+# ````
+# git clone git@github.com:trixi-framework/Trixi.jl.git 
+# # In case of an error, use following: git clone https://github.com/trixi-framework/Trixi.jl
`````
```suggestion
# # In case of an error, try the following:
# # git clone https://github.com/trixi-framework/Trixi.jl
```

sloede commented on docs/literate/src/files/getting_started.jl:
```diff
@@ -0,0 +1,288 @@
+# mkdir run 
+# cd run
+# julia --project=. -e 'using Pkg; Pkg.develop(PackageSpec(path=".."))' # Install local Trixi.jl clone
+# julia -e 'using Pkg; Pkg.add(["OrdinaryDiffEq", "Trixi2Vtk", "Plots"])' # Install additional packages'
```
```suggestion
# julia -e 'using Pkg; Pkg.add(["OrdinaryDiffEq", "Plots"])' # Install additional packages'
```

sloede commented on docs/literate/src/files/getting_started.jl:
`````diff
@@ -0,0 +1,288 @@
+# #### Linux
+
+# You can clone Trixi.jl to your computer executing following commands:
+# ````
`````
~~~suggestion
# ```shell
~~~

sloede reviewed (changes requested):
Second part of the review

ArseniyKholod commented on docs/literate/src/files/getting_started.jl:
```diff
@@ -0,0 +1,288 @@
+# And execute following code.
+
+using Trixi, OrdinaryDiffEq
+trixi_include(@__MODULE__,joinpath(examples_dir(), "tree_2d_dgsem", "elixir_euler_ec.jl"))
```
Following code runs correctly from julia text file.
`using Trixi, Plots 
trixi_include(joinpath(examples_dir(), "tree_2d_dgsem", "elixir_euler_ec.jl"))
plot(sol)
`
 But when expanding markdown templates during documentation building, `sol` is encountered as undefined:
`[ Info: ExpandTemplates: expanding markdown templates.
┌ Error: failed to run `@example` block in src\tutorials\getting_started.md:120-123
│ ```@example getting_started
│ using Plots
│ plot(sol)
│ ```
│   value = UndefVarError: sol not defined
└ @ Documenter.Expanders C:\Users\Lenovo\.julia\packages\Documenter\H5y27\src\Utilities\Utilities.jl:32`

sloede replied:
Hm. Maybe you can create [an issue with Literate.jl](https://github.com/fredrikekre/Literate.jl/issues) and ask if this is known behavior or if there is a workaround?

ArseniyKholod replied:
If instead of `trixi_include` I call `include`, this issue doesn't occur:
```
using Trixi, Plots 
include(joinpath(examples_dir(), "tree_2d_dgsem", "elixir_euler_ec.jl"))
plot(sol)
```
As I understand it, `trixi_include` in this case does exactly the same, calls `include` only with adding default `Main` module:
```
function trixi_include(mod::Module, elixir::AbstractString; kwargs...)
  # Print information on potential wait time only in non-parallel case
  if !mpi_isparallel()
    @info "You just called `trixi_include`. Julia may now compile the code, please be patient."
  end
  Base.include(ex -> replace_assignments(insert_maxiters(ex); kwargs...), mod, elixir)
end

trixi_include(elixir::AbstractString; kwargs...) = trixi_include(Main, elixir; kwargs...)
```

[The issue with Literate.jl](https://github.com/fredrikekre/Literate.jl/issues/213) is created.

ArseniyKholod replied:
Literate.jl developer [answered](https://github.com/fredrikekre/Literate.jl/issues/213#issuecomment-1529736341) on my questions.
Copy of answer:
1) Is it known behavior?
- It is known in the sense that code is not evaluated in Main (by design) but rather in a separate module.
2) Is there way around adding @MODULE?
- Not from the LIterate side. The fact that include_trixi defaults to main is either a feature or design flaw that you would have to discuss with Trixi developers. (A more intuitive behavior, IMO, would be to use a macro and default to the callers current module instead of Main, for example.)

sloede replied:
Thanks for creating that issue at Literate.jl. Is the following possible:
* **Execute** the `trixi_include(@__MODULE__, ...)` code in the tutorial but **hide** it in the Markdown file
* **Show** the `trixi_include(...)` code in the tutorial (i.e., *without* the `@__MODULE__`) but **do not execute** it?

Could this be a viable workaround?

bennibolm replied:
Yes, that's how it should work. With something like this
```
#md # ```julia
#md # using Trixi, Plots
#md # trixi_include(joinpath(examples_dir(), "tree_2d_dgsem", "elixir_euler_ec.jl"))
#md # ```

#md using Trixi, Plots #hide
#md trixi_include(@__MODULE__, joinpath(examples_dir(), "tree_2d_dgsem", "elixir_euler_ec.jl")) #hide

#!md using Trixi, Plots
#!md trixi_include(joinpath(examples_dir(), "tree_2d_dgsem", "elixir_euler_ec.jl"))

# Plot
plot(sol)
```

ArseniyKholod commented on docs/literate/src/files/getting_started.jl:
```diff
@@ -0,0 +1,288 @@
+#   the [`examples`](https://github.com/trixi-framework/Trixi.jl/tree/main/examples) folder.
+# - Navigate to the
+#   file [`elixir_euler_ec.jl`](https://github.com/trixi-framework/Trixi.jl/blob/main/examples/tree_2d_dgsem/elixir_euler_ec.jl).
+# - Click the `Raw` button on the right side of the webpage.
+# - Right-click on any place of the newly opened webpage and choose `Save as`.
+# - Choose a folder and erase `.txt` from the file name. Save the file.
```
Thanks, it's really easier!

ArseniyKholod commented on docs/literate/src/files/getting_started.jl:
```diff
@@ -0,0 +1,288 @@
+# and you will be able to download setup files from our GitHub repository, modify them,
+# and run simulations.
+
+# ## Julia installation
```
I have tried `juliaup` on Windows:
```
winget install julia -s msstore
```
It works for me only once, then I tried to uninstall and install `julia` again. Unfortunately I'm experiencing now every time a `server error`. I found many reports about this failure. Moreover there exists [way around](https://discourse.julialang.org/t/looking-for-feedback-from-users-that-currently-cant-install-juliaup-from-the-windows-store/85507) using `winget` command. But I'm not sure, that it will be long supported. So it may be better to leave that longer, but less error prone method?
I also tried `juliaup` on Xubuntu using VirtualBox:
```
curl -fsSL https://install.julialang.org | sh
```
Installation fails, but it may be caused by VirtualBox specific.

ArseniyKholod commented:
Tutorial is now ready for a review. [Preview of the first part](https://host-docs-from-fork-pr.lakemper.eu/tutorials/getting_started.html), [preview of the second part](https://host-docs-from-fork-pr.lakemper.eu/tutorials/create_first_setup.html)

bgeihe commented on docs/literate/src/files/create_first_setup.jl:
```diff
@@ -0,0 +1,256 @@
+
+stepsize_callback = StepsizeCallback(cfl = 1.6)
+
+# To save current solution in regular intervals we use [`SaveSolutionCallback`](@ref).
```
```suggestion
# To save the current solution in regular intervals we use [`SaveSolutionCallback`](@ref).
```

bgeihe commented on docs/literate/src/files/create_first_setup.jl:
```diff
@@ -0,0 +1,256 @@
+stepsize_callback = StepsizeCallback(cfl = 1.6)
+
+# To save current solution in regular intervals we use [`SaveSolutionCallback`](@ref).
+# We would like to save the initial and final solutions as well. Solution
+# will be saved as a HDF5 file located in the `out` folder. Afterwards it is possible to visualize
```
```suggestion
# We would like to save the initial and final solutions as well. The data
# will be saved as HDF5 files located in the `out` folder. Afterwards it is possible to visualize
```

bgeihe commented on docs/literate/src/files/create_first_setup.jl:
```diff
@@ -0,0 +1,256 @@
+# To save current solution in regular intervals we use [`SaveSolutionCallback`](@ref).
+# We would like to save the initial and final solutions as well. Solution
+# will be saved as a HDF5 file located in the `out` folder. Afterwards it is possible to visualize
+# a solution from saved files using Trixi2Vtk.jl and ParaView, this is described below in the
```
```suggestion
# a solution from saved files using Trixi2Vtk.jl and ParaView, which is described below in the
```

bgeihe commented on docs/literate/src/files/create_first_setup.jl:
```diff
@@ -0,0 +1,256 @@
+                                     save_final_solution = true)
+
+# Another useful callback is [`SaveRestartCallback`](@ref). It saves information for restarting
+# in regular intervals. We are interested in saving a restart file for a final solution as
```
```suggestion
# in regular intervals. We are interested in saving a restart file for the final solution as
```

bgeihe commented on docs/literate/src/files/create_first_setup.jl:
```diff
@@ -0,0 +1,256 @@
+callbacks = CallbackSet(summary_callback, analysis_callback, stepsize_callback, save_solution,
+                        save_restart)
+
+# The last step is to choose the time integration method, OrdinaryDiffEq.jl defines a wide range of
```
```suggestion
# The last step is to choose the time integration method. OrdinaryDiffEq.jl defines a wide range of
```

bgeihe commented on docs/literate/src/files/create_first_setup.jl:
```diff
@@ -0,0 +1,256 @@
+# activated at the end of each time step to perform some actions, e.g. to print a statistics.
+# We will show you how to use some of the common callbacks.
+
+# To print a summary of the simulation setup at the beginning of solve-loop
+# and to reset timers we use [`SummaryCallback`](@ref).
```
```suggestion
# To print a summary of the simulation setup at the beginning
# and to reset timers we use [`SummaryCallback`](@ref).
# When the returned callback is executed directly, the current timer values are shown.
```

bgeihe commented on docs/literate/src/files/create_first_setup.jl:
```diff
@@ -0,0 +1,256 @@
+summary_callback()
+
+# Now you can plot the solution as shown below, analyze it and improve the stability, accuracy or
+# efficiency of your setup modifying it.
```
```suggestion
# efficiency of your setup.
```

bgeihe commented on docs/literate/src/files/create_first_setup.jl:
```diff
@@ -0,0 +1,256 @@
+# ### Using Plots.jl
+
+# The first option is to use the [Plots.jl](https://github.com/JuliaPlots/Plots.jl) package
+# directly after calculations, when the solution is saved in the `sol` variable. We connect the
```
```suggestion
# directly after calculations, when the solution is saved in the `sol` variable. We load the
```

bgeihe commented on docs/literate/src/files/create_first_setup.jl:
```diff
@@ -0,0 +1,256 @@
+# [VisIt](https://visit.llnl.gov) to plot the solution. The important thing is that currently
+# Trixi2Vtk.jl supports conversion only for solutions in 2D and 3D spatial domains.
+
+# If you haven't added a Trixi2Vtk.jl to your project yet, you can add it as follows.
```
```suggestion
# If you haven't added Trixi2Vtk.jl to your project yet, you can add it as follows.
```

bgeihe commented on docs/literate/src/files/create_first_setup.jl:
````diff
@@ -0,0 +1,256 @@
+# import Pkg
+# Pkg.add(["Trixi2Vtk"])
+# ```
+# Now we are connecting the Trixi2Vtk.jl package and convert the file `out/solution_000018.h5` with
````
```suggestion
# Now we load the Trixi2Vtk.jl package and convert the file `out/solution_000018.h5` with
```

bgeihe commented on docs/literate/src/files/create_first_setup.jl:
````diff
@@ -0,0 +1,256 @@
+# Pkg.add(["Trixi2Vtk"])
+# ```
+# Now we are connecting the Trixi2Vtk.jl package and convert the file `out/solution_000018.h5` with
+# the final solution using the [`trixi2vtk`](@ref) function saving the resulted file in the
````
```suggestion
# the final solution using the [`trixi2vtk`](@ref) function saving the resulting file in the
```

bgeihe commented on docs/literate/src/files/create_first_setup.jl:
```diff
@@ -0,0 +1,256 @@
+
+# Now let's visualize the solution from the generated files in ParaView. Follow this short
+# instruction to get the visualization.
+# - Download, install and open [ParaView](https://www.paraview.org/download/)
```
```suggestion
# - Download, install and open [ParaView](https://www.paraview.org/download/).
```

bgeihe commented on docs/literate/src/files/create_first_setup.jl:
```diff
@@ -0,0 +1,256 @@
+# Now let's visualize the solution from the generated files in ParaView. Follow this short
+# instruction to get the visualization.
+# - Download, install and open [ParaView](https://www.paraview.org/download/)
+# - Press `Ctrl+O` and browse the generated files `solution_000018.vtu` and
```
```suggestion
# - Press `Ctrl+O` and select the generated files `solution_000018.vtu` and
```

bgeihe commented on docs/literate/src/files/create_first_setup.jl:
```diff
@@ -0,0 +1,256 @@
+# - In the upper-left corner in the Pipeline Browser window, left-click on the eye-icon near
+#   `solution_000018.vtu`.
+# - In the lower-left corner in the Properties window, change the Coloring from Solid Color to
+#   scalar. Now final solution visualization is already generated.
```
```suggestion
#   scalar. This already generates the visualization of the final solution.
```

bgeihe commented on docs/literate/src/files/create_first_setup.jl:
```diff
@@ -0,0 +1,256 @@
+#   scalar. Now final solution visualization is already generated.
+# - Now let's add the mesh to the visualization. In the upper-left corner in the
+#   Pipeline Browser window, left-click on the eye-icon near `solution_000018_celldata.vtu`.
+# - In the lower-left corner in the Properties window, change the Representation from the Surface
+#   to the Wireframe. Then a white grid should appear on the visualization.
```
```suggestion
# - In the lower-left corner in the Properties window, change the Representation from Surface
#   to Wireframe. Then a white grid should appear on the visualization.
```

bgeihe commented on docs/literate/src/files/create_first_setup.jl:
```diff
@@ -0,0 +1,256 @@
+# ![paraview_trixi2vtk_example](https://github.com/trixi-framework/Trixi.jl/assets/119304909/0c29139b-6c5d-4d5c-86e1-f4ebc95aca7e)
+
+
+# ## Package versions
+
+# These results were obtained using the following versions.
+
+using InteractiveUtils
+versioninfo()
+
+using Pkg
+Pkg.status(["Trixi","OrdinaryDiffEq","Plots","ForwardDiff"],
+           mode=PKGMODE_MANIFEST)
+
+Sys.rm("out"; recursive=true, force=true) #hide #md
```
Skip this, as in part 1?

bgeihe reviewed (commented):
Great to see this new introductory texts!
I hope some of my comments are helpful.

bgeihe commented on docs/literate/src/files/changing_trixi.jl:
```diff
@@ -0,0 +1,91 @@
+
+# Now you cloned Trixi.jl and only need to tell Julia to use the local clone as the package sources:
+# - Open a terminal using `Win+R` and `cmd`. Navigate to the folder with cloned Trixi.jl using `cd`.
+# - Create new directory `run`, enter it, and start Julia with the `--project=.` flag:
```
```suggestion
# - Create a new directory `run`, enter it, and start Julia with the `--project=.` flag:
```

bgeihe commented on docs/literate/src/files/changing_trixi.jl:
```diff
@@ -0,0 +1,91 @@
+
+# ### Linux
+
+# You can clone Trixi.jl to your computer executing following commands:
```
```suggestion
# You can clone Trixi.jl to your computer by executing the following commands:
```

bgeihe commented on docs/literate/src/files/changing_trixi.jl:
````diff
@@ -0,0 +1,91 @@
+# You can clone Trixi.jl to your computer executing following commands:
+# ```shell
+# git clone git@github.com:trixi-framework/Trixi.jl.git 
+# # In case of an error, try the following:
````
"case" gets highlighted. Can this be prevented?

ArseniyKholod replied:
I'm not sure, I try just do not use "case" here.

bgeihe commented on docs/literate/src/files/changing_trixi.jl:
```diff
@@ -0,0 +1,91 @@
+# mkdir run 
+# cd run
+# julia --project=. -e 'using Pkg; Pkg.develop(PackageSpec(path=".."))' # Install local Trixi.jl clone
+# julia -e 'using Pkg; Pkg.add(["OrdinaryDiffEq", "Plots"])' # Install additional packages'
```
```suggestion
# julia --project=. -e 'using Pkg; Pkg.add(["OrdinaryDiffEq", "Plots"])' # Install additional packages'
```

bgeihe commented on docs/literate/src/files/changing_trixi.jl:
```diff
@@ -0,0 +1,91 @@
+#   during a simulation, e.g., to store the solution or adapt the mesh.
+
+
+# ## Package versions
+
+# These results were obtained using the following versions.
+
+using InteractiveUtils
+versioninfo()
+
+using Pkg
+Pkg.status(["Trixi","OrdinaryDiffEq","Plots","ForwardDiff"],
+           mode=PKGMODE_MANIFEST)
+
+Sys.rm("out"; recursive=true, force=true) #hide #md
```
As before, skip this?

ArseniyKholod commented:
> Great to see this new introductory texts! I hope some of my comments are helpful.

Thank you very much! I will process out all your comments as soon as possible!

ArseniyKholod commented on docs/literate/src/files/create_first_setup.jl:
```diff
@@ -0,0 +1,256 @@
+# of elements in the mesh using `initial_refinement_level`, which describes the initial height of
+# the tree mesh. The variable `n_cells_max` is used to limit the number of elements in the mesh,
+# which cannot be exceeded in the [adaptive mesh refinement](@ref Adaptive-mesh-refinement).
+
+# All minimum and all maximum coordinates must be combined into `Tuples`.
```
As long as I understand `Tuple` is hardcoded in the constructor of `TreeMesh`
```
function TreeMesh(coordinates_min::NTuple{NDIMS, Real},
                  coordinates_max::NTuple{NDIMS, Real};
                  n_cells_max,
                  periodicity = true,
                  initial_refinement_level,
                  refinement_patches = (),
                  coarsening_patches = ()) where {NDIMS}
```
If I chane `Tuple`s to `Vector`s I got following error.
```
ERROR: LoadError: MethodError: no method matching TreeMesh(::Vector{Float64}, ::Vector{Float64}; initial_refinement_level::Int64, n_cells_max::Int64)

Closest candidates are:
  TreeMesh(::Type{TreeType}, ::Any...) where {NDIMS, TreeType<:Trixi.AbstractTree{NDIMS}} got unsupported keyword arguments "initial_refinement_level", "n_cells_max"
   @ Trixi D:\work\Trixi.jl.my\src\meshes\tree_mesh.jl:84
```

sloede replied:
OK, then just leave it as it is.

ArseniyKholod commented on docs/literate/src/files/getting_started.jl:
```diff
@@ -0,0 +1,250 @@
+    function initial_condition_density_waves(x, t, equations::CompressibleEulerEquations2D)
+        v1 = 0.1 # velocity along x-axis
+        v2 = 0.2 # velocity along y-axis
+        rho = 1.0 + 0.98 * sin(pi * (sum(x) - t * (v1 + v2))) # density wave profile
```
Thanks! It simplifies the code!

ArseniyKholod commented on docs/literate/src/files/getting_started.jl:
```diff
@@ -0,0 +1,250 @@
+# [Julia to the PATH](https://julialang.org/downloads/platform/#linux_and_freebsd).
+
+
+# ## Trixi.jl installation
```
Thank you! I believe the idea is to combine all the necessary steps for first use in one tutorial so that the newbie doesn't get lost. A pretty big part of the information in these texts has already been mentioned in other tutorials, but it's not always easy to put everything together... But maybe I'm wrong, @sloede, what would you say?

sloede replied:
Yes, my idea was that this is a one stop shop for new users, without the need to go anywhere else. Repetition is OK here IMHO

JoshuaLampert commented on docs/literate/src/files/first_steps/getting_started.jl:
```diff
@@ -0,0 +1,240 @@
+
+# Note that later on in the tutorial, Julia will be started by only typing `julia` in a terminal.
+# To enable that, you have to add
+# [Julia to the PATH](https://julialang.org/downloads/platform/#linux_and_freebsd).
```
```suggestion
# [Julia to the PATH](https://julialang.org/downloads/platform/#running_julia).
```
To point to the exact place, where this step is explained.

JoshuaLampert commented on docs/literate/src/files/first_steps/getting_started.jl:
```diff
@@ -0,0 +1,240 @@
+# happens inside Julia.
+# For a smooth workflow experience with Trixi.jl, you need to install 
+# [Trixi.jl](https://github.com/trixi-framework/Trixi.jl),
+# [OrdinaryDiffEq.jl](https://github.com/SciML/OrdinaryDiffEq.jl) and 
```
```suggestion
# [OrdinaryDiffEq.jl](https://github.com/SciML/OrdinaryDiffEq.jl), and 
```

JoshuaLampert commented on docs/literate/src/files/first_steps/getting_started.jl:
````diff
@@ -0,0 +1,240 @@
+# - Open a terminal and start Julia (Windows and Linux: type `julia`).
+# - Execute following commands:
+#   ```julia
+#   import Pkg
````
```suggestion
#   julia> import Pkg
```
Usually, we include the `julia>` prompt when writing commands that are supposed to be executed in the Julia REPL.

JoshuaLampert commented on docs/literate/src/files/first_steps/getting_started.jl:
````diff
@@ -0,0 +1,240 @@
+# - Execute following commands:
+#   ```julia
+#   import Pkg
+#   Pkg.add(["OrdinaryDiffEq", "Plots", "Trixi"])
````
```suggestion
#   julia> Pkg.add(["OrdinaryDiffEq", "Plots", "Trixi"])
```

JoshuaLampert commented on docs/literate/src/files/first_steps/getting_started.jl:
````diff
@@ -0,0 +1,240 @@
+# Start Julia in a terminal and execute the following code:
+
+# ```julia
+# using Trixi, OrdinaryDiffEq
````
```suggestion
# julia> using Trixi, OrdinaryDiffEq
```

JoshuaLampert commented on docs/literate/src/files/first_steps/getting_started.jl:
````diff
@@ -0,0 +1,240 @@
+
+# ```julia
+# using Trixi, OrdinaryDiffEq
+# trixi_include(joinpath(examples_dir(), "tree_2d_dgsem", "elixir_euler_ec.jl"))
````
```suggestion
# julia> trixi_include(joinpath(examples_dir(), "tree_2d_dgsem", "elixir_euler_ec.jl"))
```

JoshuaLampert commented on docs/literate/src/files/first_steps/getting_started.jl:
```diff
@@ -0,0 +1,240 @@
+p2 = plot(pd["v1"], clim=(0.05, 0.15)) #hide #md
+p3 = plot(pd["v2"], clim=(0.15, 0.25)) #hide #md
+p4 = plot(pd["p"], clim=(10, 30)) #hide #md
+plot(p1, p2, p3, p4) #hide #md
```
Does this give the same result as `plot(sol)`, which is suggested in the code above? If yes, why not use `plot(sol)`? If no, the user might wonder why the plot displayed in the tutorial is different than what she/he sees locally.

ArseniyKholod replied:
Thank you very much for the review!

`plot(sol)` generates a noisy picture, because 3 of four terms remain approximately constant. Such noise can also confuse user...
![image](https://github.com/trixi-framework/Trixi.jl/assets/119304909/95d3db79-0854-4d2c-bcec-56c02ff766cf)

JoshuaLampert replied:
Ok, then I suggest to also use this code also above such that the user gets the same non-noisy picture. See my suggestion.

ArseniyKholod replied:
Target audience of tutorial is people with zero experience in Julia. So all called functions are explained, so that newbie could understand the purpose of every function call. And explaining `PlotData2D` and how to configure limits for every state variable seems to be to complicated in this case... May be leave `plot(sol)`, but add a comment, that 
```
pd = PlotData2D(sol)
p1 = plot(pd["rho"])
p2 = plot(pd["v1"], clim=(0.05, 0.15))
p3 = plot(pd["v2"], clim=(0.15, 0.25))
p4 = plot(pd["p"], clim=(10, 30))
plot(p1, p2, p3, p4)
```
should be execute to get exactly the same pictures?

JoshuaLampert replied:
Yes, sounds good.

JoshuaLampert commented on docs/literate/src/files/first_steps/create_first_setup.jl:
````diff
@@ -0,0 +1,245 @@
+# \frac{\partial}{\partial t}u + \frac{\partial}{\partial x} (0.2 u) - \frac{\partial}{\partial y} (0.7 u) = - 2 e^{-t}
+# \sin\bigl(2 \pi (x - t) \bigr) \sin\bigl(2 \pi (y - t) \bigr)
+# ```
+# And with an initial condition
````
```suggestion
# and with an initial condition
```

JoshuaLampert commented on docs/literate/src/files/first_steps/create_first_setup.jl:
```diff
@@ -0,0 +1,245 @@
+# and modifying the following example.
+
+# Let's consider the linear advection equation for a state ``u = u(t, x, y)`` on the two-dimensional spatial domain
+# ``[-1, 1] \times [-1, 1]`` with a source term.
```
```suggestion
# ``[-1, 1] \times [-1, 1]`` with a source term
```

JoshuaLampert commented on docs/literate/src/files/first_steps/create_first_setup.jl:
```diff
@@ -0,0 +1,245 @@
+# [Adding a new scalar conservation law](@ref adding_new_scalar_equations) or
+# [Adding a non-conservative equation](@ref adding_nonconservative_equation).
+
+# The linear scalar advection equation is already implemented in Trixi.jl as
```
```suggestion
# The linear scalar advection equation in two spatial dimensions
# ```math
# \frac{\partial}{\partial t}u + \frac{\partial}{\partial x} (a_1 u) + \frac{\partial}{\partial y} (a_2 u) = 0
# ```
is already implemented in Trixi.jl as
```

JoshuaLampert commented on docs/literate/src/files/first_steps/create_first_setup.jl:
```diff
@@ -0,0 +1,245 @@
+
+# The linear scalar advection equation is already implemented in Trixi.jl as
+# [`LinearScalarAdvectionEquation2D`](@ref), for which we need to define a two-dimensional parameter
+# with the `advection_velocity`. Appropriate for our problem is `(0.2, -0.7)`.
```
```suggestion
# `advection_velocity` describing the paramaters ``a_1`` and ``a_2``. Appropriate for our problem is `(0.2, -0.7)`.
```

JoshuaLampert commented on docs/literate/src/files/first_steps/create_first_setup.jl:
```diff
@@ -0,0 +1,245 @@
+equations = LinearScalarAdvectionEquation2D(advection_velocity)
+
+# To solve our problem numerically using Trixi.jl, we have to discretize the spatial
+# domain, for which we set up a mesh. One of the most used meshes in Trixi.jl is
```
```suggestion
# domain, for which we set up a mesh. One of the most used meshes in Trixi.jl is the
```

JoshuaLampert commented on docs/literate/src/files/first_steps/create_first_setup.jl:
````diff
@@ -0,0 +1,245 @@
+# ```math
+# u(0,x,y) = \sin\bigl(\pi x \bigr) \sin\bigl(\pi y \bigr)
+# ```
+# The initial conditions function must take coordinates, time and equation as arguments
````
```suggestion
# The initial conditions function must take spatial coordinates, time and equation as arguments
```

JoshuaLampert commented on docs/literate/src/files/first_steps/create_first_setup.jl:
````diff
@@ -0,0 +1,245 @@
+# u(0,x,y) = \sin\bigl(\pi x \bigr) \sin\bigl(\pi y \bigr)
+# ```
+# The initial conditions function must take coordinates, time and equation as arguments
+# and return an initial condition as a static vector `SVector`. Following the same structure, you
````
```suggestion
# and returns an initial condition as a static vector `SVector`. Following the same structure, you
```

JoshuaLampert commented on docs/literate/src/files/first_steps/create_first_setup.jl:
````diff
@@ -0,0 +1,245 @@
+# ```
+# The initial conditions function must take coordinates, time and equation as arguments
+# and return an initial condition as a static vector `SVector`. Following the same structure, you
+# can define your own initial conditions.
````
Maybe add something like
```suggestion
# can define your own initial conditions. The time variable `t` can be unused in the initial condition, but might also be used to describe an analytical solution if known. If you use the initial condition as analytical solution, you can analyze your numerical solution by computing the error, see also the [section about analyzing the solution](https://trixi-framework.github.io/Trixi.jl/stable/callbacks/#Analyzing-the-numerical-solution).
```
Otherwise, a user might wonder why an initial condition needs an argument for the time. I'm also not sure if this analytical-solution-as-initial-condition behavior of Trixi.jl is properly documented somewhere.

ArseniyKholod replied:
Sounds cool!

JoshuaLampert commented on docs/literate/src/files/first_steps/create_first_setup.jl:
````diff
@@ -0,0 +1,245 @@
+
+# The next step is to define a function of the source term corresponding to our problem.
+# ```math
+# f(t,u,x,y) = - 2 e^{-t} \sin\bigl(2 \pi (x - t) \bigr) \sin\bigl(2 \pi (y - t) \bigr)
````
```suggestion
# f(u, x, y, t) = - 2 e^{-t} \sin\bigl(2 \pi (x - t) \bigr) \sin\bigl(2 \pi (y - t) \bigr)
```
Maybe change the order of arguments to reflect the order in the source term function?

JoshuaLampert commented on docs/literate/src/files/first_steps/create_first_setup.jl:
````diff
@@ -0,0 +1,245 @@
+# ```math
+# f(t,u,x,y) = - 2 e^{-t} \sin\bigl(2 \pi (x - t) \bigr) \sin\bigl(2 \pi (y - t) \bigr)
+# ```
+# This function must take the state variable, coordinates, time and the
````
```suggestion
# This function must take the state variable, the spatial coordinates, the time and the
```

JoshuaLampert commented on docs/literate/src/files/first_steps/create_first_setup.jl:
````diff
@@ -0,0 +1,245 @@
+# f(t,u,x,y) = - 2 e^{-t} \sin\bigl(2 \pi (x - t) \bigr) \sin\bigl(2 \pi (y - t) \bigr)
+# ```
+# This function must take the state variable, coordinates, time and the
+# equation itself as arguments and return the source term as a static vector `SVector`.
````
```suggestion
# equation itself as arguments and returns the source term as a static vector `SVector`.
```

JoshuaLampert commented on docs/literate/src/files/first_steps/create_first_setup.jl:
```diff
@@ -0,0 +1,245 @@
+# Since Trixi.jl has a common basic structure for the setups, you can create your own by extending
+# and modifying the following example.
+
+# Let's consider the linear advection equation for a state ``u = u(t, x, y)`` on the two-dimensional spatial domain
```
```suggestion
# Let's consider the linear advection equation for a state ``u = u(x, y, t)`` on the two-dimensional spatial domain
```
In the initial condition function further down, we use the spatial coordinates before the time variable. Being consistent between the text and the implementation in terms of the order of arguments might improve readability.

JoshuaLampert commented on docs/literate/src/files/first_steps/create_first_setup.jl:
````diff
@@ -0,0 +1,245 @@
+# ```
+# And with an initial condition
+# ```math
+# u(0,x,y) = \sin\bigl(\pi x \bigr) \sin\bigl(\pi y \bigr).
````
```suggestion
# u(x, y, 0) = \sin\bigl(\pi x \bigr) \sin\bigl(\pi y \bigr).
```

JoshuaLampert commented on docs/literate/src/files/first_steps/create_first_setup.jl:
```diff
@@ -0,0 +1,245 @@
+end
+
+# Now we are collecting all the information that will be needed to define a spatial discretization,
+# which leaves us with an ODE problem in time with a span from 0.0 s to 1.0.
```
```suggestion
# which leaves us with an ODE problem in time with a span from 0.0 to 1.0.
```
Does the unit need to be seconds?

JoshuaLampert commented on docs/literate/src/files/first_steps/create_first_setup.jl:
```diff
@@ -0,0 +1,245 @@
+ode = semidiscretize(semi, tspan);
+
+# At this point, our problem is defined. We will use the `solve` function defined in
+# OrdinaryDiffEq.jl to get the solution. OrdinaryDiffEq.jl gives us the ability to customize the solver
```
```suggestion
# [OrdinaryDiffEq.jl](https://github.com/SciML/OrdinaryDiffEq.jl) to get the solution. OrdinaryDiffEq.jl gives us the ability to customize the solver
```

JoshuaLampert commented on docs/literate/src/files/first_steps/create_first_setup.jl:
```diff
@@ -0,0 +1,245 @@
+# We will show you how to use some of the common callbacks.
+
+# To print a summary of the simulation setup at the beginning
+# and to reset timers we use [`SummaryCallback`](@ref).
```
```suggestion
# and to reset timers we use the [`SummaryCallback`](@ref).
```

JoshuaLampert commented on docs/literate/src/files/first_steps/create_first_setup.jl:
```diff
@@ -0,0 +1,245 @@
+summary_callback = SummaryCallback()
+
+# We also want to analyze the current state of the solution in regular intervals.
+# [`AnalysisCallback`](@ref) outputs some useful statistical information during the solving process
```
```suggestion
# The [`AnalysisCallback`](@ref) outputs some useful statistical information during the solving process
```

JoshuaLampert commented on docs/literate/src/files/first_steps/create_first_setup.jl:
```diff
@@ -0,0 +1,245 @@
+
+analysis_callback = AnalysisCallback(semi, interval = 5)
+
+# It is also possible to control the time step size using [`StepsizeCallback`](@ref) if the time
```
```suggestion
# It is also possible to control the time step size using the [`StepsizeCallback`](@ref) if the time
```

JoshuaLampert commented on docs/literate/src/files/first_steps/create_first_setup.jl:
```diff
@@ -0,0 +1,245 @@
+
+stepsize_callback = StepsizeCallback(cfl = 1.6)
+
+# To save the current solution in regular intervals we use [`SaveSolutionCallback`](@ref).
```
```suggestion
# To save the current solution in regular intervals we use the [`SaveSolutionCallback`](@ref).
```

JoshuaLampert commented on docs/literate/src/files/first_steps/create_first_setup.jl:
```diff
@@ -0,0 +1,245 @@
+                                     save_initial_solution = true,
+                                     save_final_solution = true)
+
+# Another useful callback is [`SaveRestartCallback`](@ref). It saves information for restarting
```
```suggestion
# Another useful callback is the [`SaveRestartCallback`](@ref). It saves information for restarting
```

JoshuaLampert commented on docs/literate/src/files/first_steps/create_first_setup.jl:
```diff
@@ -0,0 +1,245 @@
+# `CarpenterKennedy2N54(williamson_condition = false)`. We will pass the ODE
+# problem, the ODE solver and the callbacks to the `solve` function. Also, to use
+# `StepsizeCallback`, we must explicitly specify the initial trial time step `dt`, the selected
+# value is not important, because it will be overwritten by `StepsizeCallback`. And there is no
```
```suggestion
# value is not important, because it will be overwritten by the `StepsizeCallback`. And there is no
```

JoshuaLampert commented on docs/literate/src/files/first_steps/create_first_setup.jl:
````diff
@@ -0,0 +1,245 @@
+
+# If you haven't added Trixi2Vtk.jl to your project yet, you can add it as follows.
+# ```julia
+# import Pkg
````
```suggestion
# julia> import Pkg
```

JoshuaLampert commented on docs/literate/src/files/first_steps/create_first_setup.jl:
````diff
@@ -0,0 +1,245 @@
+# If you haven't added Trixi2Vtk.jl to your project yet, you can add it as follows.
+# ```julia
+# import Pkg
+# Pkg.add(["Trixi2Vtk"])
````
```suggestion
# julia> Pkg.add(["Trixi2Vtk"])
```

JoshuaLampert commented on docs/literate/src/files/first_steps/changing_trixi.jl:
```diff
@@ -0,0 +1,77 @@
+#   you want to save Trixi.jl. Then click `Clone` and Trixi.jl will be cloned to your computer. 
+
+# Now you cloned Trixi.jl and only need to tell Julia to use the local clone as the package sources:
+# - Open a terminal using `Win+R` and `cmd`. Navigate to the folder with cloned Trixi.jl using `cd`.
```
```suggestion
# - Open a terminal using `Win+R` and `cmd`. Navigate to the folder with the cloned Trixi.jl using `cd`.
```

JoshuaLampert commented on docs/literate/src/files/first_steps/changing_trixi.jl:
````diff
@@ -0,0 +1,77 @@
+#   ```
+# - Now run the following commands to install all relevant packages:
+#   ```julia
+#   using Pkg; Pkg.develop(PackageSpec(path="..")) # Install local Trixi.jl clone
````
```suggestion
#   julia> using Pkg; Pkg.develop(PackageSpec(path="..")) # Tell Julia to use the local Trixi.jl clone
```

JoshuaLampert commented on docs/literate/src/files/first_steps/changing_trixi.jl:
````diff
@@ -0,0 +1,77 @@
+# - Now run the following commands to install all relevant packages:
+#   ```julia
+#   using Pkg; Pkg.develop(PackageSpec(path="..")) # Install local Trixi.jl clone
+#   Pkg.add(["OrdinaryDiffEq", "Plots"])  # Install additional packages
````
```suggestion
#   julia> Pkg.add(["OrdinaryDiffEq", "Plots"])  # Install additional packages
```

JoshuaLampert commented on docs/literate/src/files/first_steps/changing_trixi.jl:
```diff
@@ -0,0 +1,77 @@
+# cd Trixi.jl
+# mkdir run 
+# cd run
+# julia --project=. -e 'using Pkg; Pkg.develop(PackageSpec(path=".."))' # Install local Trixi.jl clone
```
```suggestion
# julia --project=. -e 'using Pkg; Pkg.develop(PackageSpec(path=".."))' # Tell Julia to use the local Trixi.jl clone
```

JoshuaLampert commented on docs/literate/src/files/first_steps/changing_trixi.jl:
```diff
@@ -0,0 +1,77 @@
+# mkdir run 
+# cd run
+# julia --project=. -e 'using Pkg; Pkg.develop(PackageSpec(path=".."))' # Install local Trixi.jl clone
+# julia --project=. -e 'using Pkg; Pkg.add(["OrdinaryDiffEq", "Plots"])' # Install additional packages'
```
```suggestion
# julia --project=. -e 'using Pkg; Pkg.add(["OrdinaryDiffEq", "Plots"])' # Install additional packages
```

JoshuaLampert commented on docs/literate/src/files/first_steps/changing_trixi.jl:
```diff
@@ -0,0 +1,77 @@
+
+# ## For further reading
+
+# To further delve into Trixi.jl, you may have a look at following tutorials.
```
```suggestion
# To further delve into Trixi.jl, you may have a look at the following introductory tutorials.
```

JoshuaLampert reviewed (commented):
Thanks for these nice tutorials! I left some comments below.

JoshuaLampert commented on docs/literate/src/files/first_steps/getting_started.jl:
```diff
@@ -0,0 +1,240 @@
+#   using Trixi
+#   trixi_include(path/to/file)
+#   using Plots
+#   plot(sol)
```
```suggestion
#   pd = PlotData2D(sol)
#   p1 = plot(pd["rho"])
#   p2 = plot(pd["v1"], clim=(0.05, 0.15))
#   p3 = plot(pd["v2"], clim=(0.15, 0.25))
#   p4 = plot(pd["p"], clim=(10, 30))
plot(p1, p2, p3, p4)
```

ArseniyKholod commented:
Thank you very much for your reviews!
The last state of tutorial: 
https://host-docs-from-fork-pr.lakemper.eu/tutorials/first_steps/getting_started.html

JoshuaLampert commented:
Now that I'm reading it in the webversion again, I think the `julia> ` prompt should not be in front of the code lines. I'm sorry :grimacing:.

JoshuaLampert commented on docs/literate/src/files/first_steps/getting_started.jl:
```diff
@@ -0,0 +1,250 @@
+        rho_e = p / (equations.gamma - 1) + 1/2 * rho * (v1^2 + v2^2)
+        return SVector(rho, rho*v1, rho*v2, rho_e)
+    end
+    initial_condition = initial_condition_density_waves
```
```suggestion
initial_condition = initial_condition_density_waves
```
Same for the lines above. Otherwise, there is a strange looking inundation.

sloede commented:
Thanks again for this momentous tutorial! There are only three things I noticed that I think should be addressed (besides the existing comments by all the other, very helpful reviewers):
* The tutorial and a short description need to be added here: https://host-docs-from-fork-pr.lakemper.eu/tutorials/introduction.html
* The official Julia install instructions now default to installing `juliaup` (see https://julialang.org/downloads/). I think we should reflect this in the tutorial as well, otherwise it feel outdated (https://host-docs-from-fork-pr.lakemper.eu/tutorials/first_steps/getting_started.html#Julia-installation)
* In the second part (https://host-docs-from-fork-pr.lakemper.eu/tutorials/first_steps/create_first_setup.html#Using-Plots.jl), a solution plot seems to remain empty (see picture below)

Once those are fixed, I believe this is ready to merge.

<img width="881" alt="image" src="https://github.com/trixi-framework/Trixi.jl/assets/3637659/aff8b7a7-2a41-4f1e-9522-f491268744a7">

ArseniyKholod commented:
> Thanks again for this momentous tutorial! There are only three things I noticed that I think should be addressed (besides the existing comments by all the other, very helpful reviewers):
> 
> * The tutorial and a short description need to be added here: https://host-docs-from-fork-pr.lakemper.eu/tutorials/introduction.html
> * The official Julia install instructions now default to installing `juliaup` (see https://julialang.org/downloads/). I think we should reflect this in the tutorial as well, otherwise it feel outdated (https://host-docs-from-fork-pr.lakemper.eu/tutorials/first_steps/getting_started.html#Julia-installation)
> * In the second part (https://host-docs-from-fork-pr.lakemper.eu/tutorials/first_steps/create_first_setup.html#Using-Plots.jl), a solution plot seems to remain empty (see picture below)
> 
> Once those are fixed, I believe this is ready to merge.
> 
> <img alt="image" width="881" src="https://private-user-images.githubusercontent.com/3637659/302064060-aff8b7a7-2a41-4f1e-9522-f491268744a7.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDcwNDUwOTQsIm5iZiI6MTcwNzA0NDc5NCwicGF0aCI6Ii8zNjM3NjU5LzMwMjA2NDA2MC1hZmY4YjdhNy0yYTQxLTRmMWUtOTUyMi1mNDkxMjY4NzQ0YTcucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDIwNCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAyMDRUMTEwNjM0WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MzQxZmRiOTg1ZTBhY2JkYjE4NDM0YmYwNWZhZjAzYTUwMTZlMDg0NGNkOGQ2MWIzYWE0ZDUwZWZjZDJmYWEwZSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.nh5GL5cBdNTIo8_nAJpfxoMqmbHzQVeypLNb8glhJgM">

I added `juliaup` installation and short description to introduction.
Tutorial: https://host-docs-from-fork-pr.lakemper.eu/tutorials/first_steps/getting_started.html
Introduction: https://host-docs-from-fork-pr.lakemper.eu/tutorials/introduction.html

Issue with empty plot is not clear to me... Now all my devices show all pictures from tutorial, I hope it's okey now.

DanielDoehring commented on docs/literate/src/files/first_steps/changing_trixi.jl:
````diff
@@ -0,0 +1,77 @@
+# cd run
+# julia --project=. -e 'using Pkg; Pkg.develop(PackageSpec(path=".."))' # Tell Julia to use the local Trixi.jl clone
+# julia --project=. -e 'using Pkg; Pkg.add(["OrdinaryDiffEq", "Plots"])' # Install additional packages
+# ```
````
As I encountered an error here recently, we should add:

If you encounter errors when executing `julia --project=. -e 'using Pkg; Pkg.develop(PackageSpec(path=".."))'` try 
`julia --project=. -e using Pkg; Pkg.update()`

JoshuaLampert replied:
Sounds good, but should probably be `julia --project=. -e 'using Pkg; Pkg.update()'`. Note the ticks.

sloede replied:
Are you sure that this is more helpful rather than confusing for the intended target audience, i.e., Julia novices? I've never encountered this error when doing a new setup, thus I'd just omit it.

DanielDoehring replied:
I will make a seperate PR for this.

DanielDoehring replied:
See #1831

DanielDoehring commented on docs/literate/src/files/first_steps/create_first_setup.jl:
````diff
@@ -0,0 +1,258 @@
+# ```
+
+# The first step is to create and open a file with the .jl extension. You can do this with your
+# favorite text editor. In this file you will create your setup.
````
Maybe mention that VSCode is the recommended choice ?

sloede replied:
```suggestion
# favorite text editor (if you do not have one, we recommend [VS Code](https://code.visualstudio.com/)).
# In this file you will create your setup.
```

DanielDoehring commented on docs/literate/src/files/first_steps/create_first_setup.jl:
```diff
@@ -0,0 +1,258 @@
+# To solve our problem numerically using Trixi.jl, we have to discretize the spatial
+# domain, for which we set up a mesh. One of the most used meshes in Trixi.jl is the
+# [`TreeMesh`](@ref). The spatial domain used is ``[-1, 1] \times [-1, 1]``. We set an initial number
+# of elements in the mesh using `initial_refinement_level`, which describes the initial number of
+# hierarchical refinements. The variable `n_cells_max` is used to limit the number of elements in the mesh,
```
We should mention that the number of leaf cells = actual elements is computed as 2^initial_refinement_level

sloede replied:
```suggestion
# of elements in the mesh using `initial_refinement_level`, which describes the initial number of
# hierarchical refinements. In this simple case, the total number of elements is `2^initial_refinement_level`
# throughout the simulation. The variable `n_cells_max` is used to limit the number of elements in the mesh,
```

DanielDoehring commented on docs/literate/src/files/first_steps/create_first_setup.jl:
```diff
@@ -0,0 +1,258 @@
+# Now we need to define an initial condition for our problem. All the already implemented
+# initial conditions for [`LinearScalarAdvectionEquation2D`](@ref) can be found in
+# [`src/equations/linear_scalar_advection_2d.jl`](https://github.com/trixi-framework/Trixi.jl/blob/main/src/equations/linear_scalar_advection_2d.jl).
+# If you want to use, for example, a Gaussian pulse, it can be used as follows.
```
```suggestion
# If you want to use, for example, a Gaussian pulse, it can be used as follows:
```

DanielDoehring commented on docs/literate/src/files/first_steps/create_first_setup.jl:
````diff
@@ -0,0 +1,258 @@
+# u(x, y, 0) = \sin\bigl(\pi x \bigr) \sin\bigl(\pi y \bigr).
+# ```
+# The initial conditions function must take spatial coordinates, time and equation as arguments
+# and returns an initial condition as a static vector `SVector`. Following the same structure, you
````
```suggestion
# and returns an initial condition as a statically sized vector `SVector`. Following the same structure, you
```

DanielDoehring commented on docs/literate/src/files/first_steps/getting_started.jl:
```diff
@@ -0,0 +1,242 @@
+
+# ## Julia installation
+
+# Trixi.jl is compatible with the latest stable release of Julia. Additional details regarding Julia
```
Is this still true if loop vectorization gets deprecated?

sloede replied:
In general, yes, so this is ok as-is imho

DanielDoehring commented on docs/literate/src/files/first_steps/getting_started.jl:
```diff
@@ -0,0 +1,242 @@
+# +
+# \frac{\partial}{\partial x}
+# \begin{pmatrix}
+# \rho v_1 \\ \rho v_1^2 + p \\ \rho v_1 v_2 \\ (\rho e +p) v_1
```
```suggestion
# \rho v_1 \\ \rho v_1^2 + p \\ \rho v_1 v_2 \\ (\rho e + p) v_1
```

DanielDoehring replied:
```suggestion
# \rho v_1 \\ \rho v_1^2 + p \\ \rho v_1 v_2 \\ (\rho e + p) v_1
```

DanielDoehring commented on docs/literate/src/files/first_steps/getting_started.jl:
```diff
@@ -0,0 +1,242 @@
+# +
+# \frac{\partial}{\partial y}
+# \begin{pmatrix}
+# \rho v_2 \\ \rho v_1 v_2 \\ \rho v_2^2 + p \\ (\rho e +p) v_2
```
```suggestion
# \rho v_2 \\ \rho v_1 v_2 \\ \rho v_2^2 + p \\ (\rho e + p) v_2
```

DanielDoehring commented on docs/literate/src/files/first_steps/getting_started.jl:
````diff
@@ -0,0 +1,242 @@
+# Here, ``\rho`` is the density, ``v_1`` and ``v_2`` are the velocities, ``e`` is the specific
+# total energy, and
+# ```math
+# p = (\gamma - 1) \left( \rho e - \frac{1}{2} \rho (v_1^2+v_2^2) \right)
````
```suggestion
# p = (\gamma - 1) \left( \rho e - \frac{1}{2} \rho (v_1^2 + v_2^2) \right)
```

DanielDoehring reviewed (changes requested):
Nice work! I left a couple comments :)

sloede commented:
> Issue with empty plot is not clear to me... Now all my devices show all pictures from tutorial, I hope it's okey now.

Yes, thanks, works for me now too 🤷‍♂️

sloede commented on docs/literate/src/files/first_steps/getting_started.jl:
````diff
@@ -0,0 +1,242 @@
+#   ```
+#   Follow the instructions displayed in the terminal during the installation process.
+# - If an error occurs during the execution of the previous command, you may need to install
+#   `curl`. Use the following command:
````
```suggestion
#   `curl`. On Ubuntu-type systems, you can use the following command:
```

sloede reviewed (dismissed):
LGTM, thanks! once the last remaining conversation has been resolved satisfactorily, this can be merged

andrewwinters5000 commented on docs/literate/src/files/first_steps/changing_trixi.jl:
```diff
@@ -0,0 +1,77 @@
+#   you want to save Trixi.jl. Then click `Clone` and Trixi.jl will be cloned to your computer. 
+
+# Now you cloned Trixi.jl and only need to tell Julia to use the local clone as the package sources:
+# - Open a terminal using `Win+R` and `cmd`. Navigate to the folder with the cloned Trixi.jl using `cd`.
```
```suggestion
# - Open a terminal using `Win+r` and `cmd`. Navigate to the folder with the cloned Trixi.jl using `cd`.
```
In the `getting_started.jl` part of the tutorial you wrote that one should use `Win+r` for this step.

andrewwinters5000 commented on docs/literate/src/files/first_steps/changing_trixi.jl:
```diff
@@ -0,0 +1,77 @@
+# if already inside the `run` directory.
+
+
+# ## For further reading
```
```suggestion
# ## Additional reading
```

andrewwinters5000 commented on docs/literate/src/files/first_steps/create_first_setup.jl:
````diff
@@ -0,0 +1,260 @@
+# ``[-1, 1] \times [-1, 1]`` with a source term
+# ```math
+# \frac{\partial}{\partial t}u + \frac{\partial}{\partial x} (0.2 u) - \frac{\partial}{\partial y} (0.7 u) = - 2 e^{-t}
+# \sin\bigl(2 \pi (x - t) \bigr) \sin\bigl(2 \pi (y - t) \bigr)
+# ```
+# and with an initial condition
+# ```math
+# u(x, y, 0) = \sin\bigl(\pi x \bigr) \sin\bigl(\pi y \bigr).
+# ```
````
```suggestion
# \sin\bigl(2 \pi (x - t) \bigr) \sin\bigl(2 \pi (y - t) \bigr),
# ```
# with the initial condition
# ```math
# u(x, y, 0) = \sin\bigl(\pi x \bigr) \sin\bigl(\pi y \bigr),
# ```
# and periodic boundary conditions.
```

andrewwinters5000 commented on docs/literate/src/files/first_steps/create_first_setup.jl:
```diff
@@ -0,0 +1,260 @@
+    return SVector(scalar)
+end
+
+# Now we are collecting all the information that will be needed to define a spatial discretization,
```
```suggestion
# Now we collect all the information that is necessary to define a spatial discretization,
```

andrewwinters5000 commented on docs/literate/src/files/first_steps/create_first_setup.jl:
```diff
@@ -0,0 +1,260 @@
+# a solution from saved files using Trixi2Vtk.jl and ParaView, which is described below in the
+# section [Visualize the solution](@ref Visualize-the-solution).
+
+save_solution = SaveSolutionCallback(interval = 5,
+                                     save_initial_solution = true,
+                                     save_final_solution = true)
```
It might be worth mentioning here that an alternative strategy to print solution files at a fixed time interval would use the syntax
```julia
save_solution = SaveSolutionCallback(dt = 0.1,
                                     save_initial_solution = true,
                                     save_final_solution = true)
```

ArseniyKholod replied:
Thank you for the review!
Good point, I added it.

andrewwinters5000 reviewed (changes requested):
This is a very nice and thorough tutorial @ArseniyKholod ! I just left a few comments / suggestions on wording.

JoshuaLampert commented:
After resolving the merge conflicts coming from https://github.com/trixi-framework/Trixi.jl/pull/1829, this LGTM. Thanks @ArseniyKholod.

sloede commented:
@andrewwinters5000 If you believe that all your comments have been addressed satisfactorily, please feel free to merge this (you need to override auto-merge since the CI tests will never run, but the docs have been built sucessfully)

andrewwinters5000 reviewed (approved):
LGTM!
